File: jscomp/test/adt_optimize_test.ml
type t0  = A | B | C 
(* https://github.com/ocaml/ocaml/pull/632 *)


let f x  = 
  match x with 
  | B -> 2 
  | A -> 1
  | C ->  3


let f_0 x = 
  match x with 
  | A -> -1 
  | B -> 0 
  | C -> 1 

  
type t1 = 
  | T000
  | T001
  | T002
  | T003


let f2 = function  
  |'\000' -> T000
  | '\001'-> T001
  | '\002' -> T002
  | _ -> T003


type t2 = 
  | X0
  | X1 
  | X2 
  | X3 
  | X4  

type t3  =
  | Y0
  | Y1 
  | Y2 
  | Y3
  | Y4
  | Y5 of int    

let f3 = function
  | X0 -> Y0 
  | X1 -> Y1 
  | X2 -> Y2 
  | X3 -> Y3
  | X4 -> Y4 


type t4 = 
  | T400 

let f4 = function 
  | T400 -> 3  

type t5 = 
  | A 
  | B  
  | F
  | C of int 
  | D of string 
  | E of int

let f5 = function 
  | A -> 1 
  | B -> 3 
  | F -> 4
  | C _ | D _ -> 1
  | _ -> 2 

let f6 = function 
  | A | B -> 0
  | C _ | D _ | E _  ->  1 
  | _ -> 2

let f7 = function 
  | A -> 1 
  | B -> 2 
  | C _ -> 3 
  | D _ -> 4
  | _ -> -1


type t6 = 
  | T60  
  | T61
  | T62
  | T63
  | T64 of int 
  | T65 of int 
  | T66 of int 
  | T68 of int


let f8 = function 
  | T60
  | T61 -> 1 
  | T64 _ 
  | T65 _ -> 2
  | _ -> 3 


let f9 = function 
  | T60
  | T61
  | T62 -> 1 
  | T64 _ 
  | T65 _ -> 2
  | _ -> 3 

let f10 = function 
  | T60 
  | T61
  | T62 
  | T63 -> 1 
  | T64 _ 
  | T65 _ -> 2
  | _ -> 3   

let f10 = function 
  | T60 -> 0
  | T61 -> 2 
  | T62  -> 4 
  | T63 -> 1 
  | T64 _ 
  | T65 _ -> 2
  | _ -> 3     


(* let f11 = function 
  | T60 -> 0
  | T61 -> 2 
  | T62  -> 4 
  | T63 -> 1 
  | T64 _ 
  | T65 _ -> 2
  | _ -> 3        *)


type t11 = A | B | C | D of int  | E of char 

let f11 (x : t11) = 
  match x with 
  | D _ -> 1 
  | A 
  | B
  | C -> 2 

  | _ -> assert false 
  

File: jscomp/test/adt_optimize_test.js
'use strict';

var Caml_builtin_exceptions = require("../../lib/js/caml_builtin_exceptions.js");

function f(x) {
  return x + 1 | 0;
}

function f_0(x) {
  return x - 1 | 0;
}

function f2(param) {
  if (param >= 3) {
    return "T003";
  } else {
    return param;
  }
}

function f3(param) {
  return param;
}

function f4(param) {
  return 3;
}

function f5(param) {
  if (typeof param === "string") {
    switch (param) {
      case "A" :
          return 1;
      case "B" :
          return 3;
      case "F" :
          return 4;
      
    }
  } else {
    switch (/* XXX */param.tag) {
      case "C" :
      case "D" :
          return 1;
      case "E" :
          return 2;
      
    }
  }
}

function f6(param) {
  if (typeof param === "string") {
    switch (param) {
      case "A" :
      case "B" :
          return 0;
      case "F" :
          return 2;
      
    }
  } else {
    return 1;
  }
}

function f7(param) {
  if (typeof param === "string") {
    switch (param) {
      case "A" :
          return 1;
      case "B" :
          return 2;
      case "F" :
          return -1;
      
    }
  } else {
    switch (/* XXX */param.tag) {
      case "C" :
          return 3;
      case "D" :
          return 4;
      case "E" :
          return -1;
      
    }
  }
}

function f8(param) {
  if (typeof param === "string") {
    switch (param) {
      case "T60" :
      case "T61" :
          return 1;
      default:
        return 3;
    }
  } else {
    switch (/* XXX */param.tag) {
      case "T64" :
      case "T65" :
          return 2;
      default:
        return 3;
    }
  }
}

function f9(param) {
  if (typeof param === "string") {
    switch (param) {
      case "T60" :
      case "T61" :
      case "T62" :
          return 1;
      default:
        return 3;
    }
  } else {
    switch (/* XXX */param.tag) {
      case "T64" :
      case "T65" :
          return 2;
      default:
        return 3;
    }
  }
}

function f10(param) {
  if (typeof param === "string") {
    switch (param) {
      case "T60" :
          return 0;
      case "T61" :
          return 2;
      case "T62" :
          return 4;
      case "T63" :
          return 1;
      
    }
  } else {
    switch (/* XXX */param.tag) {
      case "T64" :
      case "T65" :
          return 2;
      case "T66" :
      case "T68" :
          return 3;
      
    }
  }
}

function f11(x) {
  if (typeof x === "string") {
    return 2;
  } else if (/* XXX */x.tag === "D") {
    return 1;
  } else {
    throw [
          Caml_builtin_exceptions.assert_failure,
          /* tuple */[
            "adt_optimize_test.ml",
            191,
            9
          ]
        ];
  }
}

exports.f = f;
exports.f_0 = f_0;
exports.f2 = f2;
exports.f3 = f3;
exports.f4 = f4;
exports.f5 = f5;
exports.f6 = f6;
exports.f7 = f7;
exports.f8 = f8;
exports.f9 = f9;
exports.f10 = f10;
exports.f11 = f11;
/* No side effect */
